File: samples/cxx/flamespeed2/flamespeed.cpp
/*!
 * @file flamespeed.cpp
 * C++ demo program to compute flame speeds using GRI-Mech.
 */

#include "cantera/oneD/Sim1D.h"
#include "cantera/oneD/Inlet1D.h"
#include "cantera/oneD/StFlow.h"
#include "cantera/IdealGasMix.h"
#include "cantera/transport.h"
#include <fstream>

using namespace Cantera;
using fmt::print;

int flamespeed(double phi)
{
    try {
        IdealGasMix gas("gri30_ion.xml","gri30_ion_mix");

        doublereal temp = 300.0; // K
        doublereal pressure = 1.0*OneAtm; //atm
        doublereal uin = 0.3; //m/sec

        size_t nsp = gas.nSpecies();
        vector_fp x(nsp, 0.0);

        doublereal C_atoms = 1.0;
        doublereal H_atoms = 4.0;
        doublereal ax = C_atoms + H_atoms / 4.0;
        doublereal fa_stoic = 1.0 / (4.76 * ax);
        x[gas.speciesIndex("CH4")] = 1.0;
        x[gas.speciesIndex("O2")] = 0.21 / phi / fa_stoic;
        x[gas.speciesIndex("N2")] = 0.79 / phi/ fa_stoic;

        gas.setState_TPX(temp, pressure, x.data());
        doublereal rho_in = gas.density();

        vector_fp yin(nsp);
        gas.getMassFractions(&yin[0]);

        gas.equilibrate("HP");
        vector_fp yout(nsp);
        gas.getMassFractions(&yout[0]);
        doublereal rho_out = gas.density();
        doublereal Tad = gas.temperature();
        print("phi = {}, Tad = {}\n", phi, Tad);

        //=============  build each domain ========================


        //-------- step 1: create the flow -------------

        FreeFlame flow(&gas);

        // create an initial grid
        int nz = 6;
        doublereal lz = 0.1;
        vector_fp z(nz);
        doublereal dz = lz/((doublereal)(nz-1));
        for (int iz = 0; iz < nz; iz++) {
            z[iz] = ((doublereal)iz)*dz;
        }

        flow.setupGrid(nz, &z[0]);

        // specify the objects to use to compute kinetic rates and
        // transport properties

        std::unique_ptr<Transport> trmix(newTransportMgr("Mix", &gas));
        std::unique_ptr<Transport> trmulti(newTransportMgr("Multi", &gas));

        flow.setTransport(*trmix);
        flow.setKinetics(gas);
        flow.setPressure(pressure);

        //------- step 2: create the inlet  -----------------------

        Inlet1D inlet;

        inlet.setMoleFractions(x.data());
        doublereal mdot=uin*rho_in;
        inlet.setMdot(mdot);
        inlet.setTemperature(temp);


        //------- step 3: create the outlet  ---------------------

        Outlet1D outlet;

        //=================== create the container and insert the domains =====

        std::vector<Domain1D*> domains { &inlet, &flow, &outlet };
        Sim1D flame(domains);

        //----------- Supply initial guess----------------------

        vector_fp locs{0.0, 0.3, 0.7, 1.0};
        vector_fp value;

        double uout = inlet.mdot()/rho_out;
        value = {uin, uin, uout, uout};
        flame.setInitialGuess("u",locs,value);
        value = {temp, temp, Tad, Tad};
        flame.setInitialGuess("T",locs,value);

        for (size_t i=0; i<nsp; i++) {
            value = {yin[i], yin[i], yout[i], yout[i]};
            flame.setInitialGuess(gas.speciesName(i),locs,value);
        }

        inlet.setMoleFractions(x.data());
        inlet.setMdot(mdot);
        inlet.setTemperature(temp);

        flame.showSolution();

        int flowdomain = 1;
        double ratio = 10.0;
        double slope = 0.08;
        double curve = 0.1;

        flame.setRefineCriteria(flowdomain,ratio,slope,curve);

        int loglevel=1;

        // Solve freely propagating flame

        // Linearly interpolate to find location where this temperature would
        // exist. The temperature at this location will then be fixed for
        // remainder of calculation.
        flame.setFixedTemperature(0.5 * (temp + Tad));
        flow.solveEnergyEqn();
        bool refine_grid = true;

        flame.solve(loglevel,refine_grid);
        double flameSpeed_mix = flame.value(flowdomain,flow.componentIndex("u"),0);
        print("Flame speed with mixture-averaged transport: {} m/s\n",
              flameSpeed_mix);

        // now enable ambipolar diffusion
        flow.enableAmbipolar(true);
        flame.solve(loglevel, refine_grid);
        double flameSpeed_full = flame.value(flowdomain,flow.componentIndex("u"),0);
        print("Flame speed with ambipolar diffusion: {} m/s\n",
              flameSpeed_full);

        vector_fp zvec,Tvec,HCOxvec,H3Oxvec,Evec,Uvec;

        print("\n{:9s}\t{:8s}\t{:5s}\t{:7s}\n",
              "z (m)", "T (K)", "U (m/s)", "Y(CO)");
        for (size_t n = 0; n < flow.nPoints(); n++) {
            Tvec.push_back(flame.value(flowdomain,flow.componentIndex("T"),n));
            HCOxvec.push_back(flame.value(flowdomain,flow.componentIndex("HCO+"),n));
            H3Oxvec.push_back(flame.value(flowdomain,flow.componentIndex("H3O+"),n));
            Evec.push_back(flame.value(flowdomain,flow.componentIndex("E"),n));
            Uvec.push_back(flame.value(flowdomain,flow.componentIndex("u"),n));
            zvec.push_back(flow.grid(n));
            print("{:9.6f}\t{:8.3f}\t{:5.3f}\t{:7.5f}\n",
                  flow.grid(n), Tvec[n], Uvec[n], HCOxvec[n]);
        }

        print("\nAdiabatic flame temperature from equilibrium is: {}\n", Tad);
        print("Flame speed for phi={} is {} m/s.\n", phi, Uvec[0]);

        std::ofstream outfile("flamespeed.csv", std::ios::trunc);
        outfile << "  Grid,   Temperature,   Uvec,   HCO+,    H3O+,   E\n";
        for (size_t n = 0; n < flow.nPoints(); n++) {
            print(outfile, " {:11.3e}, {:11.3e}, {:11.3e}, {:11.3e}, {:11.3e}, {:11.3e}\n",
                  flow.grid(n), Tvec[n], Uvec[n], HCOxvec[n], H3Oxvec[n], Evec[n]);
        }
    } catch (CanteraError& err) {
        std::cerr << err.what() << std::endl;
        std::cerr << "program terminating." << std::endl;
        return -1;
    }
    return 0;
}

int main()
{
    double phi;
    print("Enter phi: ");
    std::cin >> phi;
    return flamespeed(phi);
}

File: samples/matlab/flamespeed.m
%% Freely-propagating flame
%
% MATLAB demo program to compute flame speeds using GRI-Mech. This example is equivalent
% to the C++ :doc:`flamespeed.cpp <../cxx/flamespeed>` example.
%
% Requires: cantera >= 3.2.0
%
% .. tags:: Matlab, combustion, 1D flow, premixed flame, flame speed, saving output

%%
% Problem Definition
% ------------------

tic
help flamespeed

runtime = cputime;  % record the starting time

%%
% **Set parameter values**

phi = 1.;  % equivalence ratio
Tin = 300.0;  % burner temperature (K)
pressure = OneAtm;  % pressure

uIn = 0.3;  % initial guess (m/s)
lz = 0.1;  % length of simulated flow domain
nz = 6;  % initial number of grid points

logLevel = 1;
refineGrid = true;  % turn on grid refinement

%%
% **Create the gas object**
%
% This object will be used to evaluate all thermodynamic, kinetic,
% and transport properties

gas = Solution('gri30.yaml', 'gri30', 'mixture-averaged');
nsp = gas.nSpecies;

gas.setEquivalenceRatio(phi, 'CH4', 'O2:0.21,N2:0.79');
gas.TP = {Tin, pressure};
rhoIn = gas.massDensity;
Xin = gas.X;
Yin = gas.Y;

gas.equilibrate('HP');
rhoOut = gas.massDensity;
Yout = gas.Y;
Tad = gas.T;
disp(sprintf("phi = %1.1f, Tad = %1.1f\n", phi, Tad));

%%
% **Create domains required for a flame simulation**

% Create the flow domain
flame = FreeFlow(gas);
flame.setupUniformGrid(nz, lz, 0.);
flame.setSteadyTolerances(1e-5, 1e-11);
flame.P = pressure;

% Create the inlet
inlt = Inlet1D(gas);
inlt.T = Tin;
inlt.X = Xin;
inlt.massFlux = uIn * rhoIn;

% Create the outlet
outlt = Outlet1D(gas);
uOut = uIn * rhoIn / rhoOut;

%%
% **Create the stack and insert the domains**

stack = Sim1D({inlt, flame, outlt});

% Supply initial guess

locs = [0.0, 0.3, 0.7, 1.0];
flame.setProfile('velocity', locs, [uIn, uIn, uOut, uOut]);
flame.setProfile('T', locs, [Tin, Tin, Tad, Tad]);

names = gas.speciesNames;
for i = 1:gas.nSpecies
    flame.setProfile(names{i}, locs, [Yin(i), Yin(i), Yout(i), Yout(i)]);
end

stack.show();

ratio = 10.0;
slope = 0.08;
curve = 0.1;

flame.setRefineCriteria(ratio,slope,curve);

%%
% Display the initial guess and save to a container file

fprintf("Profile used for initial guess:\n\n")
flame.info

if ctUsesHDF5()
    % Cantera is compiled with native HDF5 support
    fileName = 'flamespeed.h5';
else
    fileName = 'flamespeed.yaml';
end
stack.save(fileName, 'initial-guess', 'Initial guess', true);

%%
% Solution
% --------
%
% Solve freely propagating flame

% Linearly interpolate to find location where this temperature would
% exist. The temperature at this location will then be fixed for
% remainder of calculation.
stack.setFixedTemperature(0.5 * (Tin + Tad));

%%
% **Mixture-averaged Diffusion**

flame.energyEnabled = true;
stack.solve(logLevel, refineGrid);
uVec = flame.values('velocity');
disp(sprintf("Flame speed with mixture-averaged transport: %1.4f m/s\n", uVec(1)));
stack.save(fileName, "mix", "Solution with mixture-averaged transport", true);

%%
% **Multicomponent Diffusion**

% now switch to multicomponent transport
flame.transportModel = 'multicomponent';
stack.solve(logLevel, refineGrid);
uVec = flame.values('velocity');
disp(sprintf("Flame speed with multicomponent transport: %1.4f m/s\n", uVec(1)));
stack.save(fileName, "multi", "Solution with multicomponent transport", true);

%%
% **Multicomponent Diffusion with Soret Effect**

% now enable Soret diffusion
flame.soretEnabled = true;
stack.solve(logLevel, refineGrid);
uVec = flame.values('velocity');
disp(sprintf("Flame speed with multicomponent transport + Soret: %1.4f m/s\n", uVec(1)));
stack.save(fileName, "soret", "Solution with multicomponent transport and Soret", true);

%%
% **Display solution profile**
fprintf("Profile after final solution step:\n\n")
flame.info

%%
% Show elapsed time
elapsed = cputime - runtime;
e = sprintf('Elapsed CPU time: %10.4g', elapsed);
disp(e);
toc
